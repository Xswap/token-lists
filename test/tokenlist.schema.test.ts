import Ajv from 'ajv';
import { schema } from '../src';
import exampleList from './schema/example.tokenlist.json';
import exampleNameSymbolSpecialCharacters from './schema/example-name-symbol-special-characters.tokenlist.json';
import bigExampleList from './schema/bigexample.tokenlist.json';
import exampleListMinimum from './schema/exampleminimum.tokenlist.json';
import emptyList from './schema/empty.tokenlist.json';
import bigWords from './schema/bigwords.tokenlist.json';
import invalidTokenAddress from './schema/invalidtokenaddress.tokenlist.json';
import invalidTimestamp from './schema/invalidtimestamp.tokenlist.json';
import invalidLogoURI1 from './schema/invalidlogouri.1.tokenlist.json';
import invalidLogoURI2 from './schema/invalidlogouri.2.tokenlist.json';
import invalidVersion1 from './schema/invalidversion.1.tokenlist.json';
import invalidVersion2 from './schema/invalidversion.2.tokenlist.json';
import invalidVersion3 from './schema/invalidversion.3.tokenlist.json';
import invalidDecimals1 from './schema/invaliddecimals.1.tokenlist.json';
import invalidDecimals2 from './schema/invaliddecimals.2.tokenlist.json';

const ajv = new Ajv({ allErrors: true, format: 'full' });
const validator = ajv.compile(schema);

describe('schema', () => {
  it('is valid', () => {
    expect(ajv.validateSchema(schema)).toEqual(true);
  });

  function checkSchema(schema: any, valid: boolean): void {
    expect(validator(schema)).toEqual(valid);
    expect(validator.errors).toMatchSnapshot();
  }

  it('works for example schema', () => {
    checkSchema(exampleList, true);
  });

  it('works for special characters schema', () => {
    checkSchema(exampleNameSymbolSpecialCharacters, true);
  });

  it('works for big example schema', () => {
    checkSchema(bigExampleList, true);
  });

  it('minimum example schema', () => {
    checkSchema(exampleListMinimum, true);
  });

  it('requires name, timestamp, version, tokens', () => {
    checkSchema({}, false);
  });

  it('empty list fails', () => {
    checkSchema(emptyList, false);
  });

  it('fails with big names', () => {
    checkSchema(bigWords, false);
  });

  it('checks token address', () => {
    checkSchema(invalidTokenAddress, false);
  });

  it('invalid timestamp', () => {
    checkSchema(invalidTimestamp, false);
  });

  it('invalid logo URI', () => {
    checkSchema(invalidLogoURI1, false);
    checkSchema(invalidLogoURI2, false);
  });

  it('invalid decimals', () => {
    checkSchema(invalidDecimals1, false);
    checkSchema(invalidDecimals2, false);
  });

  it('checks version', () => {
    checkSchema(invalidVersion1, false);
    checkSchema(invalidVersion2, false);
    checkSchema(invalidVersion3, false);
  });

  it('allows up to 10k tokens', () => {
    const exampleListWith10kTokens = {
      ...exampleList,
      tokens: [...Array(10000)].map(() => exampleList.tokens[0]),
    };
    checkSchema(exampleListWith10kTokens, true);
  });

  it('fails with 10001 tokens', () => {
    const exampleListWith10kTokensPlusOne = {
      ...exampleList,
      tokens: [...Array(10001)].map(() => exampleList.tokens[0]),
    };
    checkSchema(exampleListWith10kTokensPlusOne, false);
  });
});
