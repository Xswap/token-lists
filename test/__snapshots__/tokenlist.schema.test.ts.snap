// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema allows up to 10k tokens 1`] = `null`;

exports[`schema checks token address 1`] = `
Array [
  Object {
    "dataPath": ".tokens[0].address",
    "keyword": "pattern",
    "message": "should match pattern \\"^0x[a-fA-F0-9]{40}$\\"",
    "params": Object {
      "pattern": "^0x[a-fA-F0-9]{40}$",
    },
    "schemaPath": "#/properties/address/pattern",
  },
]
`;

exports[`schema checks version 1`] = `
Array [
  Object {
    "dataPath": ".version.patch",
    "keyword": "minimum",
    "message": "should be >= 0",
    "params": Object {
      "comparison": ">=",
      "exclusive": false,
      "limit": 0,
    },
    "schemaPath": "#/definitions/Version/properties/patch/minimum",
  },
]
`;

exports[`schema checks version 2`] = `
Array [
  Object {
    "dataPath": ".version.minor",
    "keyword": "minimum",
    "message": "should be >= 0",
    "params": Object {
      "comparison": ">=",
      "exclusive": false,
      "limit": 0,
    },
    "schemaPath": "#/definitions/Version/properties/minor/minimum",
  },
]
`;

exports[`schema checks version 3`] = `
Array [
  Object {
    "dataPath": ".version.major",
    "keyword": "minimum",
    "message": "should be >= 0",
    "params": Object {
      "comparison": ">=",
      "exclusive": false,
      "limit": 0,
    },
    "schemaPath": "#/definitions/Version/properties/major/minimum",
  },
]
`;

exports[`schema empty list fails 1`] = `
Array [
  Object {
    "dataPath": ".tokens",
    "keyword": "minItems",
    "message": "should NOT have fewer than 1 items",
    "params": Object {
      "limit": 1,
    },
    "schemaPath": "#/properties/tokens/minItems",
  },
]
`;

exports[`schema fails with 10001 tokens 1`] = `
Array [
  Object {
    "dataPath": ".tokens",
    "keyword": "maxItems",
    "message": "should NOT have more than 10000 items",
    "params": Object {
      "limit": 10000,
    },
    "schemaPath": "#/properties/tokens/maxItems",
  },
]
`;

exports[`schema fails with big names 1`] = `
Array [
  Object {
    "dataPath": ".tokens[0].name",
    "keyword": "maxLength",
    "message": "should NOT be longer than 40 characters",
    "params": Object {
      "limit": 40,
    },
    "schemaPath": "#/properties/name/maxLength",
  },
]
`;

exports[`schema invalid decimals 1`] = `
Array [
  Object {
    "dataPath": ".tokens[0].decimals",
    "keyword": "maximum",
    "message": "should be <= 255",
    "params": Object {
      "comparison": "<=",
      "exclusive": false,
      "limit": 255,
    },
    "schemaPath": "#/properties/decimals/maximum",
  },
]
`;

exports[`schema invalid decimals 2`] = `
Array [
  Object {
    "dataPath": ".tokens[0].decimals",
    "keyword": "minimum",
    "message": "should be >= 0",
    "params": Object {
      "comparison": ">=",
      "exclusive": false,
      "limit": 0,
    },
    "schemaPath": "#/properties/decimals/minimum",
  },
]
`;

exports[`schema invalid logo URI 1`] = `
Array [
  Object {
    "dataPath": ".tokens[0].logoURI",
    "keyword": "format",
    "message": "should match format \\"uri\\"",
    "params": Object {
      "format": "uri",
    },
    "schemaPath": "#/properties/logoURI/format",
  },
]
`;

exports[`schema invalid logo URI 2`] = `
Array [
  Object {
    "dataPath": ".logoURI",
    "keyword": "format",
    "message": "should match format \\"uri\\"",
    "params": Object {
      "format": "uri",
    },
    "schemaPath": "#/properties/logoURI/format",
  },
]
`;

exports[`schema invalid timestamp 1`] = `
Array [
  Object {
    "dataPath": ".timestamp",
    "keyword": "format",
    "message": "should match format \\"date-time\\"",
    "params": Object {
      "format": "date-time",
    },
    "schemaPath": "#/properties/timestamp/format",
  },
]
`;

exports[`schema minimum example schema 1`] = `null`;

exports[`schema requires name, timestamp, version, tokens 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'name'",
    "params": Object {
      "missingProperty": "name",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'timestamp'",
    "params": Object {
      "missingProperty": "timestamp",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'version'",
    "params": Object {
      "missingProperty": "version",
    },
    "schemaPath": "#/required",
  },
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'tokens'",
    "params": Object {
      "missingProperty": "tokens",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`schema works for big example schema 1`] = `null`;

exports[`schema works for example schema 1`] = `null`;

exports[`schema works for special characters schema 1`] = `null`;
